src/appleseed/renderer/modeling/camera/fisheyelenscamera.cpp

```diff
@@ -0,0 +1,446 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "fisheyelenscamera.h"
+
+// appleseed.renderer headers.
+#include "renderer/global/globallogger.h"
+#include "renderer/global/globaltypes.h"
+#include "renderer/kernel/rasterization/rasterizationcamera.h"
+#include "renderer/kernel/shading/shadingray.h"
+#include "renderer/modeling/camera/camera.h"
+#include "renderer/modeling/frame/frame.h"
+#include "renderer/modeling/project/project.h"
+#include "renderer/utility/transformsequence.h"
+
+// appleseed.foundation headers.
+#include "foundation/image/canvasproperties.h"
+#include "foundation/image/image.h"
+#include "foundation/math/dual.h"
+#include "foundation/math/intersection/planesegment.h"
+#include "foundation/math/matrix.h"
+#include "foundation/math/transform.h"
+#include "foundation/math/vector.h"
+#include "foundation/platform/compiler.h"
+#include "foundation/utility/api/apistring.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/autoreleaseptr.h"
+
+// Standard headers.
+#include <cstddef>
+
+// Forward declarations.
+namespace foundation    { class IAbortSwitch; }
+
+using namespace foundation;
+using namespace std;
+
+namespace renderer
+{
+
+namespace
+{
+    const char* Model = "fisheyelens_camera";
+
+    class FisheyeLensCamera
+      : public Camera
+    {
+      public:
+        FisheyeLensCamera(
+            const char*         name,
+            const ParamArray&   params)
+          : Camera(name, params)
+        {
+        }
+
+        void release() override
+        {
+            delete this;
+        }
+
+        const char* get_model() const override
+        {
+            return Model;
+        }
+
+        void print_settings() const override
+        {
+            RENDERER_LOG_INFO(
+                "camera \"%s\" settings:\n"
+                "  model                         %s\n"
+                "  film width                    %f\n"
+                "  film height                   %f\n"
+                "  focal length                  %f\n"
+                "  near-z                        %f\n"
+                "  shutter open begin time       %f\n"
+                "  shutter open end time         %f\n"
+                "  shutter close begin time      %f\n"
+                "  shutter close end time        %f\n"
+                "  projection type               %d",
```

Comment:
How can I log enum type?

Reply:
Maybe, set a string based on the enum value?